python/ray/tests/test_streaming_generator.py

```diff
@@ -799,18 +812,170 @@ async def main():
-        enable_object_reconstruction=True,
-    )
-    ray.init(address=cluster.address)
-    # Node to place the initial object.
-    node_to_kill = cluster.add_node(num_cpus=1, object_store_memory=10**8)
-    cluster.wait_for_nodes()
+def test_generator_wait(shutdown_only):
+    """
+    Make sure the generator works with ray.wait.
+    """
+    ray.init(num_cpus=8)
+
+    @ray.remote
+    def f(sleep_time):
+        for i in range(2):
+            time.sleep(sleep_time)
+            yield i
+
+    @ray.remote
+    def g(sleep_time):
+        time.sleep(sleep_time)
+        return 10
+
+    gen = f.options(num_returns="streaming").remote(1)
+
+    """
+    Test basic cases.
+    """
+    s = time.time()
+    r, ur = ray.wait([gen], num_returns=1)
+    print(time.time() - s)
+    assert len(r) == 1
+    assert ray.get(next(r[0])) == 0
+    assert len(ur) == 0
+
+    s = time.time()
+    r, ur = ray.wait([gen], num_returns=1)
+    print(time.time() - s)
+    assert len(r) == 1
+    assert ray.get(next(r[0])) == 1
+    assert len(ur) == 0
+
+    # Should raise a stop iteration.
+    for _ in range(3):
+        s = time.time()
+        r, ur = ray.wait([gen], num_returns=1)
+        print(time.time() - s)
+        assert len(r) == 1
+        with pytest.raises(StopIteration):
+            assert next(r[0]) == 0
+        assert len(ur) == 0
+
+    gen = f.options(num_returns="streaming").remote(0)
+    # Wait until the generator task finishes
+    ray.get(gen._generator_ref)
+    for i in range(2):
+        r, ur = ray.wait([gen], timeout=0)
+        assert len(r) == 1
+        assert len(ur) == 0
+        assert ray.get(next(r[0])) == i
+
+    """
+    Test the case ref is mixed with regular object ref.
+    """
+    gen = f.options(num_returns="streaming").remote(0)
+    ref = g.remote(3)
+    ready, unready = [], [gen, ref]
+    result_set = set()
+    while unready:
+        ready, unready = ray.wait(unready)
+        print(ready, unready)
+        assert len(ready) == 1
+        for r in ready:
+            if isinstance(r, StreamingObjectRefGenerator):
+                try:
+                    ref = next(r)
+                    print(ref)
+                    print(ray.get(ref))
+                    result_set.add(ray.get(ref))
+                except StopIteration:
+                    pass
+                else:
+                    unready.append(r)
+            else:
+                result_set.add(ray.get(r))
+
+    assert result_set == {0, 1, 10}
+
+    """
+    Test timeout.
+    """
+    gen = f.options(num_returns="streaming").remote(3)
+    ref = g.remote(1)
+    ready, unready = ray.wait([gen, ref], timeout=2)
+    assert len(ready) == 1
+    assert len(unready) == 1
+
+    """
+    Test num_returns
+    """
+    gen = f.options(num_returns="streaming").remote(1)
+    ref = g.remote(1)
+    ready, unready = ray.wait([ref, gen], num_returns=2)
+    assert len(ready) == 2
+    assert len(unready) == 0
+
```

Comment:
are these edge cases important:

* should have a test for `fetch_local=True` where object is larger than 100kb (through object store)? (will go through different path in core worker AFAIK)
* should have a test for `fetch_local=False`?
* is it important to test `await asyncio.wait(object_refs)` ?
* [streaming] does ray gracefully handle a generator raising an exception between yields?

Reply:
These are good test cases to add, but let's do it in a follow-up PR. I'll add these as a TODO for @rkooo567.

For the last one, it will return the exception as the last object in the generator. I think there are already tests for that in this file.

Reply:
> should have a test for fetch_local=True where object is larger than 100kb (through object store)? (will go through different path in core worker AFAIK)

Hmm, I originally thought it is just re-testing what ray.wait should have already tested. I guess there's nothing to lose by adding a simple test...

> is it important to test await asyncio.wait(object_refs) ?

This one seems unnecessary (should have already tests somewhere else)